src/Lib.hs

```diff
@@ -0,0 +1,191 @@
+
+-- | Top-level API. Examples:
+--   localhost:8081/ps
+--   localhost:8081/procps/ps?sys=aarch64-linux
+--  TODO API documentation
+type API =
+  Get '[JSON] NoContent
+    :<|> Capture "binary-name" BinaryName :> QueryParam "sys" System :> Get '[OctetStream] ByteString
+    :<|> Capture "package-name" PackageName :> Capture "binary-name" BinaryName :> QueryParam "sys" System :> Get '[OctetStream] ByteString
+
+newtype BinaryName = BinaryName {unBinaryName :: String}
+  deriving newtype (ToField, FromHttpApiData)
+
+newtype PackageName = PackageName {unPackageName :: String}
+  deriving newtype (Eq, Ord, FromField, FromHttpApiData, ToField)
+
+-- TODO just make this a string?
+-- TODO isn't this an architecture more than a system? nix calls it a system I guess
+data System
+  = X86_64_Linux
+  | AARCH_64_Linux
+
+instance Show System where
+  show X86_64_Linux = "x86_64-linux"
+  show AARCH_64_Linux = "aarch64-linux"
+
+instance FromHttpApiData System where
+  parseQueryParam "aarch64" = pure AARCH_64_Linux
+  parseQueryParam "x86_64-linux" = pure X86_64_Linux
+  parseQueryParam sys = Left $ "Unknown system architecture: " <> sys
+
+-- | A Triplet uniquely identifies a binary in nixpkgs.
+-- A Triplet is not a proof that that binary exists and is buildable, though.
+-- TODO better name for this than triplet?
+data Triplet = Triplet
+  { _tripName :: BinaryName,
+    _tripPackage :: PackageName,
+    _tripSystem :: System
+  }
+
+newtype BinariesDB = BinariesDB FilePath
+
+newtype ProgramDB = ProgramDB FilePath
+
+-- TODO make address/port configurable
+-- TODO read config parameters from the command line
+-- TODO Make nix sources configurable?
+data ServerConfig = ServerConfig
+  { _programDB :: ProgramDB,
+    _binaryDB :: BinariesDB
+  }
+
+-- | Serve the API
+-- This is a matter of
+--   1. Filling in missing package/system parameters to construct a proper 'Triplet'
+--   2. Building the triplet
+-- If an error occurs, this is returned as a 400
+server :: ServerConfig -> Server API
+server (ServerConfig programDB binDB) =
+  pure NoContent
+    :<|> (\bin -> handle bin Nothing)
+    :<|> (\pkg bin -> handle bin (Just pkg))
+  where
+    handle :: BinaryName -> Maybe PackageName -> Maybe System -> Handler ByteString
+    handle bin mpkg msys =
+      Handler . withExceptT (\err -> err400 {errBody = BSL.pack err}) $ do
+        let sys = fromMaybe X86_64_Linux msys
+        pkg <- maybe (resolvePackageName programDB bin sys) pure mpkg
+        buildTriplet binDB (Triplet bin pkg sys)
+
+-- | Check the Nix programs database for a package that provides the given binary
+-- If there is a candidate package with the same name as the binary, that package is given priority.
+-- If not, the alphabetically first package is chosen.
+resolvePackageName :: ProgramDB -> BinaryName -> System -> ExceptT String IO PackageName
+resolvePackageName (ProgramDB dbpath) bin sys = do
+  pkgs <- liftIO . withConnection dbpath $ \conn ->
```

Comment:
So we are making a connection per request.
Looks good.

Reply:
Yeah, for the `ProgramDB` I don't think it matters a lot though, but for the `ApplicationDB` there's two advantages:
- We can use `withAtomicConnection` (or whatever it's called) to make threads atomic
- If the DB disappears while running it will just make a new one. This is not something that should ever happen in prod but it's useful for testing

Reply:
At least that's how I understand it. It seems that SQLite is mostly threadsafe and can happily share connections.

Reply:
Oh right, this is for the `ProgramDB`.

Yeah, I didn't know that SQLite takes care of that much, it looks very handy.